src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+  const int prioXl = DDMinimizer::getLadderPosition(xL->second, xC->second[0]);
+  const int stairsCr = DDMinimizer::getStairCount(cR->second);
+  const int stairsXl = DDMinimizer::getStairCount(xL->second);
+  const int prioCl = DDMinimizer::getLadderPosition(cL->second, cX->second[0]);
+  const int prioXr = DDMinimizer::getLadderPosition(xR->second, cX->second[0]);
+  const int stairsCl = DDMinimizer::getStairCount(cL->second);
+  const int stairsXr = DDMinimizer::getStairCount(xR->second);
+
+  // complete case checkins and adjust the layout
+  if ((cX->second[0] != -1) &&
+      (xC->second[0] == -1 || cX->second[0] < xC->second[0])) {
+
+    if (prioCr == 0 && stairsCr > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateLeft(layout, stairsCr);
+    } else if (prioXl == 0 && stairsXl > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateRight(layout, stairsXl);
+    } else if (prioCr > 0 || prioXl > 0 || (prioCr == 0 && prioXl == 0)) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+  } else if ((xC->second[0] != -1) &&
+             (cX->second[0] == -1 || cX->second[0] > xC->second[0])) {
+
+    if ((prioCl == 0 && DDMinimizer::isFull(cL->second)) ||
+        (prioXr == 0 && DDMinimizer::isFull(xR->second))) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+
+    else if (prioCl == 0 && stairsCl > 0) {
+      layout = DDMinimizer::rotateRight(layout, stairsCl);
+    } else if (prioXr == 0 && stairsXr > 0) {
+      layout = DDMinimizer::rotateLeft(layout, stairsXr);
+    }
+  } else if ((DDMinimizer::isFull(xR->second) ||
+              DDMinimizer::isFull(cL->second))) {
+    layout = DDMinimizer::reverseLayout(layout);
+  } else {
+    return DDMinimizer::createControlBasedPermutation(qc);
+  }
+
+  // transform layout into permutation
+  qc::Permutation perm;
+  std::vector<Qubit> layer(bits);
+  for (qc::Qubit i = 0; i < bits; i++) {
+    layer[i] = i;
+  }
+  for (qc::Qubit i = 0; i < bits; i++) {
+    perm[layer[i]] = layout[i];
+  }
+  return perm;
+}
+
+// Helper function to create the data structure for the pattern analysis of
+// controlled gates
+std::pair<std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>>,
+          std::map<std::string, std::vector<int>>>
+DDMinimizer::makeDataStructure(qc::QuantumComputation& qc) {
+  const std::size_t bits = qc.getNqubits();
+
+  // map string of cX, xC, cL, xL, cR, xR to map of control and target bit
+  // to the index of the gate
+  std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>> maps;
+
+  // string of cX, xC, cL, xL, cR, xR to vector of max index of gate
+  // for cL and xL position 0 in the vector marks the line of c(x) at 0 -> we
+  // count the left most as the first for cR and xR position 0 in the vector
+  // marks the line of c(x) at bits - 1 -> we count the right most as the first
+  std::map<std::string, std::vector<int>> indices;
```

Comment:
If you know the exact keys of your map beforehand, using a map does not seem to be the most efficient solution. Could you imagine defining the six maps contained in `maps` as separate variables? If you really want to group them into one object, you could define a structure with six members corresponding to those six maps. However, I think the latter is not necessary here.

Btw, `std::map` is an ordered map which means that its implementation is quite inefficient. Whenever you are using a map, you most likely want to use `std::unordered_map` which is an efficient implementation of a hash map. When you do that, do not forget to adapt the includes accordingly, i.e., add `#include <unordered_map>`.